HW1/P6/P6.py

```diff
@@ -0,0 +1,68 @@
+from pyspark import SparkContext
+import numpy as np
+sc =SparkContext()
+
+nPartitions = 20
+# Read in files
+data = sc.textFile("shakespear.txt").collect()
+
+rdd = sc.parallelize(data, nPartitions).flatMap(lambda x: x.split())
+# Function that filters out words as required.
+def contextFilter(rdd):
+	rdd = rdd.filter(lambda x: x.isdigit() == False)
+	rdd = rdd.filter(lambda x: x.isupper() == False)
+	rdd = rdd.filter(lambda x: x[:-1].isupper() == False)
+	return rdd
+
+rdd = contextFilter(rdd)
+
+# For each word, add its first and second lag.
+def concatenate(rdd):
+	words1 = rdd.collect()
+	words2 = words1[1:] + ["/"]
+	words3 = words2[1:] + ["/"]
+
+	rdd1 = sc.parallelize(words1, nPartitions).zipWithIndex().map(lambda x: (x[1], x[0]))
+	rdd2 = sc.parallelize(words2, nPartitions).zipWithIndex().map(lambda x: (x[1], x[0]))
+	rdd3 = sc.parallelize(words3, nPartitions).zipWithIndex().map(lambda x: (x[1], x[0]))
+	rdd = rdd1.union(rdd2).union(rdd3)
+	return rdd
+
+rdd = concatenate(rdd)
+
+# Place the word into the format of ((word1, word2), word3)
+def placeTuple(x):
+	elem = tuple(x[1])
+	return ((elem[0], elem[1]), elem[2])
+
+rdd = rdd.groupByKey().map(placeTuple)
```

Comment:
I am not sure whether it is safe (keep the order as wanted) to union and then group-by-key. I may suggest joinByKey.